File: solutions/cpp/0133.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;

    Node() {}

    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (!node) return NULL;

        if (!ans.count(node)) {
            ans[node] = new Node(node->val, {});
            for (Node* neighbor : node->neighbors)
                ans[node]->neighbors.push_back(cloneGraph(neighbor));
        }

        return ans[node];
    }

private:
    unordered_map<Node*, Node*> ans;
};

File: solutions/python3/0133.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, neighbors):
        self.val = val
        self.neighbors = neighbors
"""


class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return None

        if node not in self.ans:
            self.ans[node] = Node(node.val, [])
            for neighbor in node.neighbors:
                self.ans[node].neighbors.append(self.cloneGraph(neighbor))

        return self.ans[node]

    ans = {}
